### PR TITLE
Added ALLOW_NEGATIVE_REBASE_REPORTING env variable

### PR DESCRIPTION
**Pull request by skhomuti**
It uses new consensus check_sanity() function before report sending

chore: Split simulate rebase function into cl and el rebases

### CODE REVIEW

File: src/modules/accounting/accounting.py [hunk 1]
```diff
@@ -93,6 +96,16 @@ def is_extra_data_submitted(self, blockstamp: BlockStamp) -> bool:
         processing_state = self._get_processing_state(blockstamp)
         return processing_state.extra_data_submitted
 
+    def check_sanity(self, blockstamp: ReferenceBlockStamp) -> bool:
+        if ALLOW_NEGATIVE_REBASE_REPORTING:
+            return True
+        cl_rebase_report = self.simulate_cl_rebase(blockstamp)
+        frame_cl_rebase = self.bunker_service.get_cl_rebase_for_current_report(blockstamp, cl_rebase_report)
+        if frame_cl_rebase < 0:
+            logger.info({'msg': 'CL rebase is negative.', 'value': frame_cl_rebase})
```
Comment (avsetsin) [inline src/modules/accounting/accounting.py:105]: I would suggest showing this message as a warning on several lines and even if `ALLOW_NEGATIVE_REBASE_REPORTING` is turned on

File: src/modules/submodules/consensus.py [hunk 2]
```diff
@@ -199,6 +199,9 @@ def process_report(self, blockstamp: ReferenceBlockStamp) -> None:
 
         report_hash = self._get_report_hash(report_data)
         logger.info({'msg': 'Calculate report hash.', 'value': report_hash})
+        if not self.check_sanity(blockstamp):
+            logger.info({'msg': 'Sanity check is not passed. Report will not be sent.'})
```
Comment (avsetsin) [inline src/modules/submodules/consensus.py:203]: Let's shout about it

**Review by avsetsin — APPROVED** [review]

File: src/modules/ejector/ejector.py [hunk 3]
```diff
@@ -265,3 +265,6 @@ def is_main_data_submitted(self, blockstamp: BlockStamp) -> bool:
 
     def is_contract_reportable(self, blockstamp: BlockStamp) -> bool:
         return not self.is_main_data_submitted(blockstamp)
+
+    def check_sanity(self, blockstamp: BlockStamp) -> bool:
```
Comment (F4ever) [inline src/modules/ejector/ejector.py:269]: Add comment why it always True

File: src/modules/submodules/consensus.py [hunk 4]
```diff
@@ -352,3 +355,7 @@ def is_main_data_submitted(self, blockstamp: BlockStamp) -> bool:
     @abstractmethod
     def is_contract_reportable(self, blockstamp: BlockStamp) -> bool:
         """Returns true if contract is ready for report"""
+
+    @abstractmethod
+    def check_sanity(self, blockstamp: ReferenceBlockStamp) -> bool:
+        """Check if collected report data is valid."""
```
Comment (F4ever) [inline src/modules/submodules/consensus.py:361]: The data is valid, but the build_output is unexpected and need to be checked manually

File: src/variables.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 # - App specific -
 LIDO_LOCATOR_ADDRESS = os.getenv('LIDO_LOCATOR_ADDRESS')
 FINALIZATION_BATCH_MAX_REQUEST_COUNT = os.getenv('FINALIZATION_BATCH_MAX_REQUEST_COUNT', 1000)
+ALLOW_NEGATIVE_REBASE_REPORTING = bool(int(os.getenv('ALLOW_NEGATIVE_REBASE_REPORTING', 0)))
```
Comment (F4ever) [inline src/variables.py:19]: Maybe better this way?
```suggestion
ALLOW_NEGATIVE_REBASE_REPORTING = os.getenv('ALLOW_NEGATIVE_REBASE_REPORTING', 'False') == 'True'
```

File: src/modules/accounting/accounting.py [hunk 6]
```diff
@@ -93,6 +96,18 @@ def is_extra_data_submitted(self, blockstamp: BlockStamp) -> bool:
...
+        cl_rebase_report = self.simulate_cl_rebase(blockstamp)
+        frame_cl_rebase = self.bunker_service.get_cl_rebase_for_current_report(blockstamp, cl_rebase_report)
+        if frame_cl_rebase < 0:
+            logger.warning({'msg': '!' * 50})
+            logger.warning({'msg': 'CL rebase is negative.', 'value': frame_cl_rebase})
+            logger.warning({'msg': '!' * 50})
+            if ALLOW_NEGATIVE_REBASE_REPORTING:
+                return True
+            return False
+        return True
```
Comment (F4ever) [inline src/modules/accounting/accounting.py:109]:
```suggestion
        if frame_cl_rebase >= 0:
            retrun True
            
       logger.warning({'msg': '!' * 50})
       logger.warning({'msg': 'CL rebase is negative.', 'value': frame_cl_rebase})
       logger.warning({'msg': '!' * 50})
       return ALLOW_NEGATIVE_REBASE_REPORTING:
```

File: src/modules/accounting/accounting.py [hunk 7]
```diff
@@ -190,15 +205,22 @@ def _get_withdrawal_batches(self, blockstamp: ReferenceBlockStamp) -> list[int]:
...
+        simulation = self.simulate_el_rebase(blockstamp)
         shares_rate = simulation.post_total_pooled_ether * SHARE_RATE_PRECISION_E27 // simulation.post_total_shares
         logger.info({'msg': 'Calculate shares rate.', 'value': shares_rate})
         return shares_rate
 
-    def get_rebase_after_report(
+    def simulate_cl_rebase(self, blockstamp: ReferenceBlockStamp) -> LidoReportRebase:
+        return self.simulate_rebase_after_report(blockstamp)
+
+    def simulate_el_rebase(self, blockstamp: ReferenceBlockStamp) -> LidoReportRebase:
```
Comment (F4ever) [inline src/modules/accounting/accounting.py:216]: Not el, but full rebase

File: src/modules/submodules/consensus.py [hunk 8]
```diff
@@ -199,6 +199,9 @@ def process_report(self, blockstamp: ReferenceBlockStamp) -> None:
 
         report_hash = self._get_report_hash(report_data)
         logger.info({'msg': 'Calculate report hash.', 'value': report_hash})
+        if not self.check_sanity(blockstamp):
```
Comment (F4ever) [inline src/modules/submodules/consensus.py:202]: Add comment what is sanity

File: src/modules/accounting/accounting.py (hunk 6 above)
Comment (skhomuti) [reply src/modules/accounting/accounting.py:109]: yay, that's better

File: src/modules/submodules/consensus.py (hunk 8 above)
Comment (skhomuti) [reply src/modules/submodules/consensus.py:204]: added, look at this

File: src/modules/accounting/accounting.py (hunk 7 above)
Comment (skhomuti) [reply src/modules/accounting/accounting.py]: renamed

File: src/variables.py (hunk 5 above)
Comment (skhomuti) [reply src/variables.py]: yea, agree

File: src/modules/submodules/consensus.py (hunk 4 above)
Comment (skhomuti) [reply src/modules/submodules/consensus.py]: rephrased

File: src/modules/ejector/ejector.py (hunk 3 above)
Comment (skhomuti) [reply src/modules/ejector/ejector.py:269]: added

File: README.md [hunk 9]
```diff
@@ -50,13 +50,14 @@ docker build -t lidofinance/oracle .
...
-| `MEMBER_PRIV_KEY`      | Private key of the Oracle member account | False    | `0x1...`                |
+| Name                              | Description                                                  | Required | Example value             |
+|-----------------------------------|--------------------------------------------------------------|----------|---------------------------|
+| `EXECUTION_CLIENT_URI`            | URI of the Execution Layer client                            | True     | `http://localhost:8545`   |
+| `CONSENSUS_CLIENT_URI`            | URI of the Consensus Layer client                            | True     | `http://localhost:5052`   |
+| `KEYS_API_URI`                    | URI of the Keys API                                          | True     | `http://localhost:8080`   |
+| `LIDO_LOCATOR_ADDRESS`            | Address of the Lido contract                                 | True     | `0x1...`                  |
+| `ALLOW_NEGATIVE_REBASE_REPORTING` | If 0, a report with negative cl rebase would not be reported | False    | `true`, `false` (default) |
```
Comment (F4ever) [inline README.md:59]:
```suggestion
| `ALLOW_NEGATIVE_REBASE_REPORTING` | If 'False', a report with negative cl rebase would not be reported | False    | `True` |
```

**Pull request merged**